src/core/exec-invoke.c

```diff
@@ -2548,6 +2557,157 @@ static int create_many_symlinks(const char *root, const char *source, char **sym
+                .dqb_bhardlimit = block_limit,
+                .dqb_bsoftlimit = 0,
+                .dqb_ihardlimit = inode_limit,
+                .dqb_isoftlimit = 0,
+                .dqb_valid = QIF_LIMITS
+        };
+
+        r = RET_NERRNO(quotactl_fd(fd, QCMD_FIXED(Q_SETQUOTA, PRJQUOTA), proj_id, &req));
+        if (r < 0)
+                return r;
+
+        log_debug("Storage quotas set for project id %u. Block limit = %" PRIu64 ", inode limit = %" PRIu64, proj_id, block_limit, inode_limit);
+
+        return r;
+}
+
+static int apply_exec_quotas(const char *target_dir, const char *cgroup_path, ExecDirectoryType type, const QuotaLimit *ql) {
+
+        if (!IN_SET(type, EXEC_DIRECTORY_STATE, EXEC_DIRECTORY_CACHE, EXEC_DIRECTORY_LOGS))
+                return 0;
+
+        assert(target_dir);
+
+        _cleanup_close_ int fd = -EBADF;
+        int quota_supported = 0;
+        int r = 0;
+
+        fd = xopenat(AT_FDCWD, target_dir, O_RDONLY|O_CLOEXEC|O_NOCTTY|O_NOFOLLOW);
+        if (fd < 0)
+                return log_debug_errno(fd, "Exec quotas: Failed to open %s: %m", target_dir);
+
+        uint32_t proj_id = 0;
+        r = get_proj_id(fd, &proj_id);
+        if (r < 0)
+                return log_debug_errno(r, "Exec quotas: Failed to retrieve project id for %s: %m", target_dir);
+
+        bool proj_id_exists = PROJ_ID_MIN <= proj_id && proj_id <= PROJ_ID_MAX;
+
+        if (!ql->quota_enforce) {
+                if (proj_id_exists) {
+                        struct dqblk req;
+                        quota_supported = is_proj_id_quota_supported(fd, proj_id, &req);
+                        if (quota_supported < 0)
+                                return log_debug_errno(quota_supported, "Exec quotas: Failed to query disk quota for project id %u: %m", proj_id);
+
+                        /* Do not enforce quotas anymore */
+                        if (quota_supported && FLAGS_SET(req.dqb_valid, QIF_BLIMITS) && (req.dqb_bhardlimit > 0 || req.dqb_ihardlimit > 0)) {
+                                req.dqb_bhardlimit = 0;
+                                req.dqb_ihardlimit = 0;
+
+                                r = RET_NERRNO(quotactl_fd(fd, QCMD_FIXED(Q_SETQUOTA, PRJQUOTA), proj_id, &req));
+                                if (r < 0)
+                                        return log_debug_errno(r, "Exec quotas: Failed to disable project quotas for %s: %m", target_dir);
+
+                                log_debug("Storage quotas for %s with project id %u were disabled", target_dir, proj_id);
+                        }
+
+                        /* Release project id if no accounting needed */
+                        if (!ql->quota_accounting) {
+                                r = set_proj_id_recursive(target_dir, 0);
+                                if (r < 0)
+                                        log_warning_errno(r, "Exec quotas: Failed to release project id %u for %s: %m", proj_id, target_dir);
```

Comment:
we have this rule in systemd that whenever we log about an error we ignore we suffix the message with ", ignoring", to indicate that to the user. hence here and above: `…"Failed to release project ID %" PRIu32 " for '%s', ignoring: %m"…`. you get the idea.